### ISSUE TITLE
[New rule] Notify that foreach may throw InvalidCastException

### ISSUE THREAD

**Issue by maxkoshevoi** [dotnet/runtime#54629]
- Continuation of: https://github.com/dotnet/runtime/issues/48529 (that issue got locked)
- Analyzer description: https://github.com/dotnet/roslyn-analyzers/issues/4479

I tried to move it to StyleCop, but it got rejected there as well: https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/3307#issuecomment-785952062 (as they don't deal with analyzers that prevent exception in any way. only style related ones)

I agree that including this in roslyn as a `warning` will break a lot of existing code, but I still feel like this analyzer will bring value to .Net community, so it shouldn't be rejected outright.

**False positives**

There was a lot of concern about false positive rate, but after some investigation, I believe it can be greatly reduced.

Regarding what is and isn't a false positive, there are too two groups so far:
- `object` collections (like `ArrayList`). A lot of code is already written for them and it expects `foreach` to cast them to some type, so there shouldn't be any surprise that it performs explicit cast in their case. I excluded such collections from this analyzer, so this is no longer a false positive.
  I'm not sure if generic collections with `object` as type argument (`List<object>`) should be included in this bucket though. Would like to hear any thoughts on this.
- Collections that always return one type. For example:
  ```cs
  foreach (string item in GenerateSequence()) { }
  
  IEnumerable<IComparable> GenerateSequence() { }
  ```
  Here's an example from `dotnet/runtime`. [Here](https://github.com/dotnet/runtime/blob/master/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaType.cs#L226) explicit cast is performed since `Instantiation` [returns `TypeDesc`](https://github.com/dotnet/runtime/blob/main/src/coreclr/tools/Common/TypeSystem/Common/Instantiation.cs#L98) and `GenericParameterDesc` is derived from `TypeDesc`. Should it be flagged with this analyzer to manually perform the cast? Also would like to hear any thoughts on this. [Here](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSchemas.cs#L212) is another example (`schema.Includes` is `XmlSchemaObject`)

As for other false positives, I didn't find any so far (at least in `dotnet/runtime` nothing is flagged if I exclude everything mentioned above).

**Proposal**

I propose to include this analyzer into roslyn, but as a `suggestion` (or at least `disabled`, if you'll come to conclusion that is still will be too noisy). So it's available to all users out of the box.

People seemed to be kinda ok with `suggestion` severity:
- https://github.com/dotnet/roslyn-analyzers/pull/4868#discussion_r579463692
- https://github.com/dotnet/runtime/issues/48529#issuecomment-782430301

CC: @jmarolf, @stephentoub, @AraHaan, @GrabYourPitchforks, @mavasani

**Comment by stephentoub** [comment]
My opinion on this hasn't changed since all of my feedback on the previous issue; I don't think it's a rule we should be adding to the SDK.

**Comment by jmarolf** [comment]
@maxkoshevoi I think the best way forward here is to have you create an analyzer in one of the existing DotNetAnalyzers repos such as https://github.com/DotNetAnalyzers/VarUsageAnalyzer. this can be released to nuget.org and folks can adopt it if they want. If you have an army of folks that all use this analyzer and find it valuable then we can revisit the decision of including it as an option in roslyn or the SDK as seems appropriate.

**Comment by Kir-Antipov** [comment]
Let me give my opinion on this rule, since I strongly believe this is not something we may have, but absolutely **must** have

`foreach` does one weird thing: it makes explicit cast implicit. Back in the days, in pre-generic era, it was kinda fine, since there was no way to tell the compiler that our collection utilizes objects of some given type. But nowadays this behavior is drastically outdated and does more harm than good. There's a reason we can't use constructions like this:
```c#
object obj = ...;
string str = obj;
```
There's the reason explicit casts are explicit. It makes developers ~reevaluate their life choices,~ rethink what they're doing, make sure the types are indeed compatible. It reduces the influence of inattention on the code. 

`foreach` **is not** the way of writing the required cast. `foreach` is the way of iterating over some object. It's not casting operator, is it?

So, once again,

```c#
object obj = "1";
int i = obj;
```

and 

```c#
object[] objects = new object[] { "1" };
foreach (int i in objects)
{
    // ...
}
```

are **the same** thing, but one is forbidden, and another is not. There's the curse of backward compatibility, but we can and must make sure such code will never ever become a thing again.

---

Few words about "false positives" and the need to ignore:

> I tried the rule from your PR on dotnet/runtime, and it yielded 1199 warnings; I sampled them, and none of the ones I looked at were bugs.

Of course, they aren't. 0% of bear attack survivors died. Existing code isn't really good measure for such things. 

> Given the high false positive rate, this rule if approved, should most likely be a disabled by default rule OR silent/hidden so it’s a refactoring.

Just like other warnings, this one shouldn't be muted, but fixed. The analyzer should provide 3 fixes:

Example to be refactored:
```c#
IEnumerable<object> objects = ...;
foreach (int i in objects)
{
    Console.WriteLine(i);
}
```

1) Make the cast explicit
```c#
foreach (object uncastedI in objects)
{
    int i = (object)uncastedI;
    Console.WriteLine(i);
}
```
This option may be used as a quick fix, since it will produce the same IL in release build as the original code. No warnings, the logic persists, the cast is explicit (as it should be) and intentional, and everyone's happy

2) Use `.Cast<T>`
```c#
foreach (int i in objects.Cast<int>())
{
    Console.WriteLine(i);
}
```
Another way to make the cast explicit is the `System.Linq.Enumerable.Cast` method

3) Use `.OfType<T>`
```c#
foreach (int i in objects.OfType<int>())
{
    Console.WriteLine(i);
}
```
My personal favorite, since `InvalidCastException` is not an option anymore :)

---

TL;DR: implicit explicit cast performed by `foreach` was not that bad back in the days, but now it should be avoided. False positives in existing (and, *obviously*, working just fine) projects cannot be a reason to reject the tool that can help people avoid unobvios InvalidCastException in the future. 1199 warnings can be fixed by one quick fix and one commit, see no problems there.

**Comment by stephentoub** [comment]
As @jmarolf outlined,  you are of course welcome to write an analyzer and make it available on nuget for anyone who wants it to use,  whether in your own library or an existing one.

**Comment by Kir-Antipov** [comment]
@stephentoub, why should it be a separate nuget? This one perfectly makes sense as a built-in rule, it's not opinion-based like `VarUsageAnalyzer`

**Comment by AraHaan** [comment]
I agree this is more or so a way to notify people that their code can cause more problems than good.

**Comment by stephentoub** [comment]
> it's not opinion-based

Sure it is.  Per your example, it's saying that the syntax:
```C#
foreach (object o in objects)
{
    int i = (int)o;
    ...
}
```
is somehow better than the syntax:
```C#
foreach (int i in object)
{
    ...
}
```
and that the language's long-standing support for expressing the cast in that simple manner is somehow inferior to being more explicit about it.  That is an opinion.

**Comment by mavasani** [comment]
Agree with @stephentoub. .NET SDK analyzers are targeted towards improving code quality, either based on better .NET API usage or generic code quality improvement. There is absolutely no code quality improvement from this proposed rule. The rule either belongs to a third party analyzer package, or may be a candidate for [code style rules](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/). If you wish to pursue the latter, please port the issue to https://github.com/dotnet/roslyn/issues so the Roslyn IDE/code style team can triage it. Either way, we should close this issue.

**Comment by Kir-Antipov** [comment]
> Sure it is

@stephentoub, I don't get it.

Facts:
 - `foreach` is an **iteration statement**, it's responsible for iterating over objects. Due to SRP, it has nothing to do with casting
 - implicit cast to derived type is not a thing in C#
 - implicit explicit casting was introduced as a workaround, since there was no other simple way to work with non-generic collections
 - generic is now a thing
 - LINQ is now a thing

Conclusion:
- There's no need in dirty hacks, that can cause additional problems, anymore

I don't see an "opinion" here.

---

> Per your example, it's saying that the syntax: [...] is somehow better than the syntax: [...]

I'm not saying it's better (I prefer `.OfType`, as I said before), it's just one of the possibilities to avoid that C# 1.0 workaround (imagine that we don't have it now). Since it produces the same IL, I propose it as a default quick fix for legacy code.

We can argue about the way to express explicit casting, this part is really opinion-based, but the main idea itself is not.

**Comment by KalleOlaviNiemitalo** [comment]
A few special cases to consider if you intend to implement a code fixer as part of the analyzer.

> Use `.Cast<T>`

This seems likely to cause additional memory allocations that cost performance:

- possibly boxing the original enumerable and/or enumerator
- possibly boxing each non-null item in order to return them from the non-generic IEnumerator.Current
- allocating the LINQ IEnumerator\<T>

And won't be possible at all if the collection has a GetEnumerator method but does not implement IEnumerable, like Span\<T>.

> Use `.OfType<T>`

If T is a reference type (or nullable value type?), then this will discard null elements, unlike the original foreach loop.

**Comment by Kir-Antipov** [comment]
@KalleOlaviNiemitalo, true. That's why first fix is an option :)

**Comment by maxkoshevoi** [comment]
@jmarolf It's too late to break people's code over it, but what about suggestion or disabled rule that can be turned on?

Also, as mentioned by @mavasani, this rule may not even be included in SDK, but in IDE for example (like a style suggestion). 
Reject came from `runtime` team, so I just wanted to first reopen issue here, before doing so in `roslyn`.

**Comment by jmarolf** [comment]
We can discuss it over on https://github.com/dotnet/roslyn but I generally agree with @stephentoub. If no one has asked for this (or manually used this style) in 20 years  its unclear to my why we should add it now.

**Comment by maxkoshevoi** [comment]
> If no one has asked for this

Well, not no one:

- Here's a stackoverflow question about this (4k views): https://stackoverflow.com/questions/949798/foreach-can-throw-an-invalidcastexception/949835
- Unity forum question: https://forum.unity.com/threads/solved-invalid-cast-error-foreach-loop.878938/

- Existing warning in Rider: https://www.jetbrains.com/help/rider/PossibleInvalidCastExceptionInForeachLoop.html
- Another existing analyzer: https://rules.sonarsource.com/csharp/RSPEC-3217

But yeah, lets move this into `roslyn`. 
@jmarolf Could you reopen this issue please: https://github.com/dotnet/roslyn-analyzers/issues/4479

**Comment by jmarolf** [comment]
> @jmarolf Could you reopen this issue please: dotnet/roslyn-analyzers/issues/4479

re-opened and transferred

**Comment by Kir-Antipov** [comment]
> foreach has implicitly casted since it was first introduced 20 years ago. All documentation we've ever had has made note of this.

Exactly what I've just said. 

> If this was 20 years ago we could discuss warning people about this case but it just doesn't make sense to do so today.

Those were different times. As I said before, C# 1.0 had its reasons to *(kind of)* justify this thing. But it's not wine or something, it doesn't get better with time. Every single use case of the implicit explicit cast was obliterated with time, now it's just a bad practice people should be aware of

**Comment by jmarolf** [comment]
@Kir-Antipov if you think this is something that should change, I would start a discussion on [dotnet/csharplang](https://github.com/dotnet/csharplang/discussions/new).

My point is this is how the language has always worked and people are just as easily confused by analyzers that warn them:
- https://youtrack.jetbrains.com/issue/RSRP-407286
- https://youtrack.jetbrains.com/issue/RSRP-427294
- https://youtrack.jetbrains.com/issue/RSRP-339122
- https://youtrack.jetbrains.com/issue/RSRP-297669
- https://youtrack.jetbrains.com/issue/RSRP-296595
- https://youtrack.jetbrains.com/issue/RSRP-297636
- https://youtrack.jetbrains.com/issue/RSRP-297633
- https://youtrack.jetbrains.com/issue/RSRP-297622
- https://youtrack.jetbrains.com/issue/RSRP-297632

**Comment by Kir-Antipov** [comment]
@jmarolf, I don't think it should :) It will be a breaking change and such things ain't welcome by C#. So there's no need to bother language designers

In this particular case I see the analyzer as a tool that can change something without actually breaking the language

**Comment by carlossanlop** [comment]
Closing this issue in favor of https://github.com/dotnet/roslyn/issues/54712
